src/ray/common/ray_config.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/ray_config.h"
+
+#include <sstream>
+#include <typeinfo>
+
+#include "nlohmann/json.hpp"
+#include "ray/util/logging.h"
+
+using json = nlohmann::json;
+
+RayConfig &RayConfig::instance() {
+  static RayConfig config;
+  return config;
+}
+
+/// -----------Include ray_config_def.h to set config items.-------------------
+/// A helper macro that helps to set a value to a config item.
+#define RAY_CONFIG(type, name, default_value) \
+  if (pair.key() == #name) {                  \
+    name##_ = pair.value().get<type>();       \
+    continue;                                 \
+  }
+
+void RayConfig::initialize(const std::string &config_list) {
+  if (config_list.empty()) {
+    return;
+  }
+  try {
+    // Parse the configuration list.
+    json config_map = json::parse(config_list);
+    for (const auto &pair : config_map.items()) {
+      // We use a big chain of if else statements because C++ doesn't allow
+      // switch statements on strings.
+#include "ray/common/ray_config_def.h"
+      RAY_LOG(FATAL) << "Received unexpected config parameter " << pair.key();
```

Comment:
Can you comment about this line? I am a little confused why we have that

Reply:
The Ray cluster won't start if we passed a non-existent config key, e.g. there's a typo or the key is outdated and has been removed recently.

PS: This is the old logic. I moved it from `ray_config.h` to `ray_config.cc`.